src/fpm_backend_output.f90

```diff
@@ -0,0 +1,182 @@
+!># Build Backend Progress Output
+!> This module provides a derived type `build_progress_t` for printing build status
+!> and progress messages to the console while the backend is building the package.
+!>
+!> The `build_progress_t` type supports two modes: `normal` and `plain`
+!> where the former does 'pretty' output and the latter does not.
+!> The `normal` mode is intended for typical interactive usage whereas
+!> 'plain' mode is used with the `--verbose` flag or when `stdout` is not attached
+!> to a terminal (e.g. when piping or redirecting `stdout`). In these cases,
+!> the pretty output must be suppressed to avoid control codes being output.
+
+module fpm_backend_output
+use iso_fortran_env, only: stdout=>output_unit
+use fpm_filesystem, only: basename
+use fpm_targets, only: build_target_ptr
+use fpm_backend_console, only: console_t
+use M_attr, only: attr, attr_mode
+implicit none
+
+private
+public build_progress_t
+
+!> Build progress object
+type build_progress_t
+    !> Console object for updating console lines
+    type(console_t) :: console
+    !> Number of completed targets
+    integer :: n_complete
+    !> Total number of targets scheduled
+    integer :: n_target
+    !> 'Plain' output (no colors or updating)
+    logical :: plain_mode = .true.
+    !> Store needed when updating previous console lines
+    integer, allocatable :: output_lines(:)
+    !> Queue of scheduled build targets
+    type(build_target_ptr), pointer :: target_queue(:)
+contains
+    !> Initialise build progress object
+    procedure :: init => output_init

```

Comment:
Constructors shouldn't be type-bound, since this makes inheritance difficult/impossible. I recommend to either use a separate procedure with `type(...), intent(out)` or overload the constructor interface `build_progress_t`.